x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -482,6 +487,27 @@ private static List<NamedExpression> combineProjections(
             return replaced;
         }
 
+        private static List<Expression> combineUpperGroupingsAndLowerProjections(
```

Comment:
This method looks quite similar to combineProjections()  which is the reused inside projectAggregations().
Have you looked into reusing them on groupings?

Reply:
I initially wanted to re-use this, but I need to make some assumptions (in the form of casting expressions) that are only true if the upper expressions are groupings. I can probably still combine these, but that might become confusing to read.